services/pull/pull.go

```diff
@@ -57,6 +57,40 @@ func NewPullRequest(repo *models.Repository, pull *models.Issue, labelIDs []int6
 
 	notification.NotifyNewPullRequest(pr)
 
+	// add first push codes comment
+	baseGitRepo, err := git.OpenRepository(pr.BaseRepo.RepoPath())
+	if err != nil {
+		return err
+	}
+	defer baseGitRepo.Close()
+
+	compareInfo, err := baseGitRepo.GetCompareInfo(pr.BaseRepo.RepoPath(),
+		pr.BaseBranch, pr.GetGitRefName())
+	if err != nil {
+		return err
+	}
+
+	if compareInfo.Commits.Len() > 0 {
+		comitIDs := ""
+		for e := compareInfo.Commits.Front(); e != nil; e = e.Next() {
+			commitID := e.Value.(*git.Commit).ID.String()
+			comitIDs = commitID + ":" + comitIDs
+		}
+
+		comitIDs = comitIDs[0 : len(comitIDs)-1]
+
+		ops := &models.CreateCommentOptions{
+			Type:        models.CommentTypePullPush,
+			Doer:        pull.Poster,
+			Repo:        repo,
+			Issue:       pr.Issue,
+			IsForcePush: false,
+			Content:     comitIDs,
```

Comment:
If I'm not mistaken, confused or too tired 😉, the problem here is that a force-push will make some commits disappear from the repository (maybe not immediately, but after garbage collection). Then the comment will fail to display anything or produce a 500 Server Error. That's why I proposed the json solution: we'd be taking a "snapshot" of the current state and the comment would always be valid. Can you confirm if this scenario can be a problem or not?

Reply:
I don't think so, because if it is a force-push, IIt willn't try to get commit message. only show ids. see:
https://github.com/a1012112796/gitea/blob/e3481b06dc562b43af0ddfaa40617f482fed96fe/models/issue_comment.go#L560-L573

Reply:
By the way, I have try "Garbage collect all repositories" in admin Maintenance Operations today, 
But I found It hven't remove commits that are not on  branch. I don't  known which thing this function did...

Reply:
The git-gc command has two levels of "aggressiveness"; I don't know which one does Gitea. There are [more options](https://git-scm.com/docs/git-gc), too. 🤷‍♂️